src/components/Layout/DataTypes.tsx

```diff
@@ -0,0 +1,283 @@
+      return (
+        <>
+          {value.map((v, i) => (
+            <DataTypeLine
+              key={i}
+              label={i === 0 ? label : ''}
+              depth={depth}
+              dataType={{ ...Contains, Value: v }}
+            />
+          ))}
+        </>
+      )
+
+    case Type.Object:
+      const { Name, Children } = dataType
+
+      if (Name === 'Cid') {
+        const cid = Children['/'].Value as string
+        return (
+          <DataTypeLine
+            label={`${label} (CID)`}
+            depth={depth}
+            dataType={{ Name, Type: Type.String, Value: cid }}
+          />
+        )
+      }
+
+      return (
+        <DataTypeMapLines label={label} depth={depth} dataTypeMap={Children} />
+      )
+
+    default:
+      throw new Error(`Unexpected DataType: ${JSON.stringify(dataType)}`)
+  }
+}
+
+interface DataTypeLinesProps {
+  label?: string
+  depth?: number
+  dataType: DataType
+}
+
+DataTypeLines.propTypes = {
+  label: PropTypes.string,
+  depth: PropTypes.number,
+  dataType: PropTypes.object.isRequired
+}
+
+/**
+ * DataTypeMapLines
+ * Renders a DataTypeMap on multiple lines
+ */
+
+export const DataTypeMapLines = ({
+  label,
+  depth,
+  dataTypeMap
+}: DataTypeMapLinesProps) => (
+  <>
+    {label && <Line label={label} depth={depth} />}
+    {Object.entries(dataTypeMap).map(([key, dt]) => (
+      <DataTypeLines
+        key={key}
+        label={key}
+        depth={label ? (depth ?? 0) + 1 : depth}
+        dataType={dt}
+      />
+    ))}
+  </>
+)
+
+interface DataTypeMapLinesProps {
+  label?: string
+  depth?: number
+  dataTypeMap: DataTypeMap
+}
+
+DataTypeMapLines.propTypes = {
+  label: PropTypes.string,
+  depth: PropTypes.number,
+  dataTypeMap: PropTypes.object.isRequired
+}
+
+/**
+ * DataTypeLine
+ * Renders the value of a DataType on a single line
+ */
+
+export const DataTypeLine = ({ label, depth, dataType }: DataTypeLineProps) => (
+  <Line label={label} depth={depth}>
+    <DataTypeValue dataType={dataType} />
+  </Line>
+)
+
+interface DataTypeLineProps {
+  label?: string
+  depth?: number
+  dataType: DataType
+}
+
+DataTypeLine.propTypes = {
+  label: PropTypes.string,
+  depth: PropTypes.number,
+  dataType: PropTypes.object.isRequired
+}
+
+/**
+ * DataTypeValue
+ * Renders the value of a DataType using the most suitable component.
+ */
+
+export const DataTypeValue = ({ dataType }: DataTypeValueProps) => {
+  const { Name, Value } = dataType
+
+  switch (dataType.Type) {
+    case Type.Bool:
+      const boolVal = Value as boolean
+      return <>{boolVal.toString()}</>
+
+    case Type.String:
+      const strVal = Value as string
+
+      if (Name === 'Address')
+        return (
+          <AddressLink address={strVal} fetchAddress hideCopyText={false} />
+        )
+
+      if (Name === 'FilecoinNumber')
+        return <>{new FilecoinNumber(strVal, 'attofil').toFil() + ' FIL'}</>
+
+      return <>{Value}</>
+
+    case Type.Number:
```

Comment:
I think all Number types in the VM use 18 decimal precision, so they'll _all_ be in the `atto` SI prefix unit, meaning we should still be able to 

```
return <>{new FilecoinNumber(strVal, 'attofil').toFil()}
```

And just leave it without the "FIL" denominator at the end like the FilecoinNumber type has

Reply:
I'm not sure if I understand correctly. When `dataType.Type === Type.Number`, it should just be a regular number, like a transaction ID or something, but not related to Filecoin values. Then it should always be a string.

Reply:
Ahh ok got it. I thought you were using Number for anything that wasn't specifically a Filecoin balance